### ISSUE TITLE
Saving two checkpoints instead of one (+ worse BLEU)

### ISSUE THREAD

**Issue by martinpopel** [tensorflow/tensor2tensor#556]
T2T had switched from schedule `train_and_evaluate` to `continuous_train_and_eval` and since some version (I think it was v1.4.0) it saves by default two checkpoints instead of one after each 2000 steps. So by default the checkpoints are saved at steps 1, 2000, 2001, 4000, 4001, 6000, 6001 etc.
This has been reported as one of the problems in #495, but now I found even more strange observation. When evaluating each checkpoint (with `t2t-bleu`, i.e. independently on the internal approx_bleu evaluation), I found that the *001 checkpoints are much worse than the *000 checkpoints (except for 2000 and 2001):
![two-checkpoints-bug](https://user-images.githubusercontent.com/724617/35784375-fb8c8a4c-0a16-11e8-9b74-02e177de8348.png)

**Comment by martinpopel** [comment]
I ran two experiments (with optimizer=Adafactor, warmup=8k, T2T v1.4.2, eight GPUs) which **differ only in the `--schedule` parameter**: `train` (gray) vs. `continuous_train_and_eval` (green)
![bug-continous-train-and-eval-vs-train](https://user-images.githubusercontent.com/724617/35870800-dcaa1b4c-0b62-11e8-9f7a-bf109a72c0ba.png)
Both experiments are evaluated with `t2t-bleu`.
For the `train` (gray) experiment there is a checkpoint each hour.
For the `continuous_train_and_eval` (green) there are two checkpoints each 2000 steps (each 45 minutes).
The `train` experiment does no internal evaluation, thus it is a bit faster (14560 steps in 4 hours, versus 12000 steps in 4 hours for the `continuous_train_and_eval`), but this is not my point now (and I choose steps instead of time for the x-axis, so the graph is not influenced by this).
Obviously, my point that the `continuous_train_and_eval` curve is much worse even if we ignore the sudden drop to almost zero BLEU at 6k and 8k steps.

I believe this may be related to the drops and double checkpoins reported in my previous post. I was not able to test the `train_and_evaluate` schedule because it fails on OOM error (which is expected, I think it needs to store two models in the memory at one time).

**Comment by rsepassi** [comment]
Wow, very odd.

My hypothesis is that it has something to do with how it goes through the data. With `continuous_train_and_eval` it recalls the `input_fn` for training each time it switches back to it. It won't see the data as evenly as `schedule=train`.

**Comment by martinpopel** [comment]
Now I am reading more carefully the [continuous_train_and_eval documentation](https://www.tensorflow.org/api_docs/python/tf/contrib/learn/Experiment#continuous_train_and_eval), which says

> THIS FUNCTION IS EXPERIMENTAL.

I had thought this warning is outdated (as it is the default in T2T now)... before creating this issue.

> First, the resources (e.g., memory) used by training will be released before evaluation (`train_and_evaluate` takes double resources).

This is great, we save memory.

> Second, more checkpoints will be saved as a checkpoint is generated at the end of each training iteration.

This explains the double checkpoints are to be expected. I know when stopping the training (with whatever schedule) e.g. at step 1000 and resuming from this checkpoint, the first thing T2T does is saving checkpoint 1001. I am not sure if it actually does the one step (training one minibatch) before saving the checkpoint, or not and increasing the step count is just done to prevent overwriting the old checkpoint.

> As the estimator.train starts from scratch (new graph, new states for input, etc) at each iteration, it is recommended to have the train_steps_per_iteration larger. It is also recommended to shuffle your input.

Does it mean that after each internal evaluation the training starts really from scratch, i.e. from the first minibatch? Thus effectively training only on the first 2000 minibatches all the time? I can't believe it. But if not, why they suggest to shuffle the input (after each internal evaluation)?

@rsepassi: When I resume training from a checkpoint, does the training starts from the very first training example, or does it compute the position in the data (based on global_step and batch_size and number of GPUs) where it had ended last time and continues from there?

**Comment by rsepassi** [comment]
There is no logic to keep track of where training was the last time it stopped. The input function is called from scratch again. However, I don't think it's as bad as starting from the exact same first minibatch because we do shuffle the filenames and the data as we read it.

**Comment by rsepassi** [comment]
And my guess is that 1 minibatch of training is done for step 1001. And because that minibatch may have already been seen (or at least it's not as uncorrelated as we'd like it to be) it slightly degrades performance.

**Comment by martinpopel** [comment]
> I don't think it's as bad as starting from the exact same first minibatch because we do shuffle the filenames

By default there are just 100 data shards (filenames), so shuffling the filenames *only* would mean training on 100 * 2000 minibatches...

> and the data as we read it.

... but I guess this means that it starts from a random position in the file, am I right? (I saw the code, but thought the shuffling is there only to prevent bottleneck disk access/deadlocks in distributed training.)

I guess this shuffling is influenced by the random seed and I hope the seed is not reset after each evaluation.
It is reset/fixed only when starting `t2t-trainer`. Thus if I kill `t2t-trainer` always after it saves the first checkpoint (after 2000 steps of training) and resume, I will get the problem of training on the same 2000 minibatches. Am I right?
(I just want to understand how it works. I don't say this is a bug - hopefully nobody needs to kill&resume the training so often. Personally, I use `--schedule=train`, so I don't care about this issue at all, I just wanted to report because most users use the default schedule.)

**Comment by martinpopel** [comment]
The error is still there in T2T 1.5.2.

**Comment by songmeixu** [comment]
> > I don't think it's as bad as starting from the exact same first minibatch because we do shuffle the filenames
> 
> By default there are just 100 data shards (filenames), so shuffling the filenames _only_ would mean training on 100 * 2000 minibatches...
> 
> > and the data as we read it.
> 
> ... but I guess this means that it starts from a random position in the file, am I right? (I saw the code, but thought the shuffling is there only to prevent bottleneck disk access/deadlocks in distributed training.)
> 
> I guess this shuffling is influenced by the random seed and I hope the seed is not reset after each evaluation.
> It is reset/fixed only when starting `t2t-trainer`. Thus if I kill `t2t-trainer` always after it saves the first checkpoint (after 2000 steps of training) and resume, I will get the problem of training on the same 2000 minibatches. Am I right?
> (I just want to understand how it works. I don't say this is a bug - hopefully nobody needs to kill&resume the training so often. Personally, I use `--schedule=train`, so I don't care about this issue at all, I just wanted to report because most users use the default schedule.)

Hi, @martinpopel , your information is very valuable for me. As to your question here, namely does the training see only a part of whole data (of course in `--schedule=continuous_train_and_eval`)?

I'm also wondering it, do you have an answer now?
